chain/client/src/sync/state/mod.rs

```diff
@@ -0,0 +1,388 @@
+                    *num_concurrent_requests_during_catchup
+                } else {
+                    *num_concurrent_requests
+                } as usize;
+                let fallback_source = Arc::new(StateSyncDownloadSourceExternal {
+                    clock: clock.clone(),
+                    store: store.clone(),
+                    chain_id: chain_id.to_string(),
+                    conn: external,
+                    timeout,
+                }) as Arc<dyn StateSyncDownloadSource>;
+                (
+                    Some(fallback_source),
+                    *external_storage_fallback_threshold as usize,
+                    num_concurrent_requests.min(NUM_CONCURRENT_REQUESTS_FOR_PEERS),
+                )
+            } else {
+                (None, 0, NUM_CONCURRENT_REQUESTS_FOR_PEERS)
+            };
+
+        let (header_validation_sender, header_validation_queue) =
+            tokio::sync::mpsc::unbounded_channel();
+        let (chain_finalization_sender, chain_finalization_queue) =
+            tokio::sync::mpsc::unbounded_channel();
+
+        let downloading_task_tracker = TaskTracker::new(num_concurrent_requests);
+        let downloader = Arc::new(StateSyncDownloader {
+            clock,
+            store: store.clone(),
+            preferred_source: peer_source,
+            fallback_source,
+            num_attempts_before_fallback,
+            header_validation_queue: header_validation_sender,
+            runtime: runtime.clone(),
+            retry_timeout: timeout, // TODO: This is not what timeout meant. Introduce a new parameter.
+            task_tracker: downloading_task_tracker.clone(),
+        });
+
+        let num_concurrent_computations = if catchup {
+            NUM_CONCURRENT_REQUESTS_FOR_COMPUTATION_DURING_CATCHUP
+        } else {
+            NUM_CONCURRENT_REQUESTS_FOR_COMPUTATION
+        };
+        let computation_task_tracker = TaskTracker::new(num_concurrent_computations);
+
+        Self {
+            store,
+            peer_source_state,
+            downloader,
+            downloading_task_tracker,
+            computation_task_tracker,
+            future_spawner,
+            epoch_manager,
+            runtime,
+            header_validation_queue,
+            chain_finalization_queue,
+            chain_finalization_sender,
+            shard_syncs: HashMap::new(),
+        }
+    }
+
+    /// Apply a state sync message received from a peer.
+    pub fn apply_peer_message(
+        &self,
+        peer_id: PeerId,
+        shard_id: ShardId,
+        sync_hash: CryptoHash,
+        data: ShardStateSyncResponse,
+    ) -> Result<(), near_chain::Error> {
+        self.peer_source_state
+            .lock()
+            .unwrap()
+            .receive_peer_message(peer_id, shard_id, sync_hash, data)?;
+        Ok(())
+    }
+
+    /// Processes the requests that the state sync module needed the Chain for.
+    fn process_chain_requests(&mut self, chain: &mut Chain) {
```

Comment:
Slightly wary about having this setup for sync stuff happening...
While this is objectively better for our use case here, we have a pattern where we send an actix message to client to handle all things that should be in sync in client and this sorta breaks that... I can't think of better ways or alternatives

Reply:
Changed to AsyncSender.